File: packages/sdk-python/src/clawtoken/contracts.py
"""Contracts API — create, sign, fund, milestones, disputes, settlement."""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

from clawtoken.http import AsyncHttpClient, HttpClient


class ContractsApi:
    """Synchronous Contracts API."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def list(self, **params: Any) -> dict[str, Any]:
        """List contracts (with optional filters)."""
        return self._http.get("/api/contracts", params or None)

    def get(self, contract_id: str) -> dict[str, Any]:
        """Get contract by ID."""
        return self._http.get(f"/api/contracts/{quote(contract_id, safe='')}")

    def create(self, **kwargs: Any) -> dict[str, Any]:
        """Create a new service contract."""
        return self._http.post("/api/contracts", kwargs)

    def sign(self, contract_id: str, **kwargs: Any) -> dict[str, Any]:
        """Sign a contract."""
        return self._http.post(f"/api/contracts/{quote(contract_id, safe='')}/sign", kwargs)

    def fund(self, contract_id: str, **kwargs: Any) -> dict[str, Any]:
        """Fund a contract."""
        return self._http.post(f"/api/contracts/{quote(contract_id, safe='')}/fund", kwargs)

    def complete(self, contract_id: str, **kwargs: Any) -> dict[str, Any]:
        """Complete a contract."""
        return self._http.post(f"/api/contracts/{quote(contract_id, safe='')}/complete", kwargs)

    # -- Milestones ----------------------------------------------------------

    def submit_milestone(self, contract_id: str, milestone_id: str, **kwargs: Any) -> dict[str, Any]:
        """Submit a milestone deliverable."""
        return self._http.post(
            f"/api/contracts/{quote(contract_id, safe='')}/milestones/{quote(milestone_id, safe='')}/submit",
            kwargs,
        )

    def approve_milestone(self, contract_id: str, milestone_id: str, **kwargs: Any) -> dict[str, Any]:
        """Approve a submitted milestone."""
        return self._http.post(
            f"/api/contracts/{quote(contract_id, safe='')}/milestones/{quote(milestone_id, safe='')}/approve",
            kwargs,
        )

    def reject_milestone(self, contract_id: str, milestone_id: str, **kwargs: Any) -> dict[str, Any]:
        """Reject a submitted milestone."""
        return self._http.post(
            f"/api/contracts/{quote(contract_id, safe='')}/milestones/{quote(milestone_id, safe='')}/reject",
            kwargs,
        )

    # -- Disputes ------------------------------------------------------------

    def open_dispute(self, contract_id: str, **kwargs: Any) -> dict[str, Any]:
        """Open a dispute on a contract."""
        return self._http.post(f"/api/contracts/{quote(contract_id, safe='')}/dispute", kwargs)

    def resolve_dispute(self, contract_id: str, **kwargs: Any) -> dict[str, Any]:
        """Resolve a contract dispute."""
        return self._http.post(f"/api/contracts/{quote(contract_id, safe='')}/dispute/resolve", kwargs)

    # -- Settlement ----------------------------------------------------------

    def settlement(self, contract_id: str, **kwargs: Any) -> dict[str, Any]:
        """Execute final settlement."""
        return self._http.post(f"/api/contracts/{quote(contract_id, safe='')}/settlement", kwargs)


class AsyncContractsApi:
    """Asynchronous Contracts API."""

    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    async def list(self, **params: Any) -> dict[str, Any]:
        return await self._http.get("/api/contracts", params or None)

    async def get(self, contract_id: str) -> dict[str, Any]:
        return await self._http.get(f"/api/contracts/{quote(contract_id, safe='')}")

    async def create(self, **kwargs: Any) -> dict[str, Any]:
        return await self._http.post("/api/contracts", kwargs)

    async def sign(self, contract_id: str, **kwargs: Any) -> dict[str, Any]:
        return await self._http.post(f"/api/contracts/{quote(contract_id, safe='')}/sign", kwargs)

    async def fund(self, contract_id: str, **kwargs: Any) -> dict[str, Any]:
        return await self._http.post(f"/api/contracts/{quote(contract_id, safe='')}/fund", kwargs)

    async def complete(self, contract_id: str, **kwargs: Any) -> dict[str, Any]:
        return await self._http.post(f"/api/contracts/{quote(contract_id, safe='')}/complete", kwargs)

    async def submit_milestone(self, contract_id: str, milestone_id: str, **kwargs: Any) -> dict[str, Any]:
        return await self._http.post(
            f"/api/contracts/{quote(contract_id, safe='')}/milestones/{quote(milestone_id, safe='')}/submit",
            kwargs,
        )

    async def approve_milestone(self, contract_id: str, milestone_id: str, **kwargs: Any) -> dict[str, Any]:
        return await self._http.post(
            f"/api/contracts/{quote(contract_id, safe='')}/milestones/{quote(milestone_id, safe='')}/approve",
            kwargs,
        )

    async def reject_milestone(self, contract_id: str, milestone_id: str, **kwargs: Any) -> dict[str, Any]:
        return await self._http.post(
            f"/api/contracts/{quote(contract_id, safe='')}/milestones/{quote(milestone_id, safe='')}/reject",
            kwargs,
        )

    async def open_dispute(self, contract_id: str, **kwargs: Any) -> dict[str, Any]:
        return await self._http.post(f"/api/contracts/{quote(contract_id, safe='')}/dispute", kwargs)

    async def resolve_dispute(self, contract_id: str, **kwargs: Any) -> dict[str, Any]:
        return await self._http.post(f"/api/contracts/{quote(contract_id, safe='')}/dispute/resolve", kwargs)

    async def settlement(self, contract_id: str, **kwargs: Any) -> dict[str, Any]:
        return await self._http.post(f"/api/contracts/{quote(contract_id, safe='')}/settlement", kwargs)

File: packages/sdk/src/contracts.ts
/**
 * Contracts API — create, sign, fund, milestones, disputes, settlement.
 */
import type { HttpClient, RequestOptions } from './http.js';
import type {
  Contract,
  CreateContractParams,
  CreateContractResponse,
  ContractActionParams,
  ContractFundParams,
  MilestoneSubmitParams,
  MilestoneApproveParams,
  MilestoneRejectParams,
  ContractDisputeParams,
  ContractDisputeResolveParams,
  ContractSettlementParams,
} from './types.js';

interface TxHashResponse {
  txHash: string;
  [key: string]: unknown;
}

export class ContractsApi {
  constructor(private readonly http: HttpClient) {}

  /** List contracts (optionally filtered by status or party). */
  async list(
    params?: { status?: string; party?: string; limit?: number; offset?: number },
    opts?: RequestOptions,
  ): Promise<{ contracts: Contract[]; total: number }> {
    return this.http.get<{ contracts: Contract[]; total: number }>(
      '/api/v1/contracts',
      params as Record<string, string | number>,
      opts,
    );
  }

  /** Get contract by ID. */
  async get(contractId: string, opts?: RequestOptions): Promise<Contract> {
    return this.http.get<Contract>(`/api/v1/contracts/${enc(contractId)}`, undefined, opts);
  }

  /** Create a new service contract. */
  async create(
    params: CreateContractParams,
    opts?: RequestOptions,
  ): Promise<CreateContractResponse> {
    return this.http.post<CreateContractResponse>('/api/v1/contracts', params, opts);
  }

  /** Sign a contract. */
  async sign(
    contractId: string,
    params: ContractActionParams,
    opts?: RequestOptions,
  ): Promise<TxHashResponse> {
    return this.http.post<TxHashResponse>(
      `/api/v1/contracts/${enc(contractId)}/actions/sign`,
      params,
      opts,
    );
  }

  /** Fund a contract (lock escrow). */
  async fund(
    contractId: string,
    params: ContractFundParams,
    opts?: RequestOptions,
  ): Promise<TxHashResponse> {
    return this.http.post<TxHashResponse>(
      `/api/v1/contracts/${enc(contractId)}/actions/activate`,
      params,
      opts,
    );
  }

  /** Mark contract as completed. */
  async complete(
    contractId: string,
    params: ContractActionParams,
    opts?: RequestOptions,
  ): Promise<TxHashResponse> {
    return this.http.post<TxHashResponse>(
      `/api/v1/contracts/${enc(contractId)}/actions/complete`,
      params,
      opts,
    );
  }

  // -------------------------------------------------------------------------
  // Milestones
  // -------------------------------------------------------------------------

  /** Submit a milestone deliverable. */
  async submitMilestone(
    contractId: string,
    milestoneId: string,
    params: MilestoneSubmitParams,
    opts?: RequestOptions,
  ): Promise<TxHashResponse> {
    return this.http.post<TxHashResponse>(
      `/api/v1/contracts/${enc(contractId)}/milestones/${enc(milestoneId)}/actions/submit`,
      params,
      opts,
    );
  }

  /** Approve a submitted milestone. */
  async approveMilestone(
    contractId: string,
    milestoneId: string,
    params: MilestoneApproveParams,
    opts?: RequestOptions,
  ): Promise<TxHashResponse> {
    return this.http.post<TxHashResponse>(
      `/api/v1/contracts/${enc(contractId)}/milestones/${enc(milestoneId)}/actions/approve`,
      params,
      opts,
    );
  }

  /** Reject a submitted milestone. */
  async rejectMilestone(
    contractId: string,
    milestoneId: string,
    params: MilestoneRejectParams,
    opts?: RequestOptions,
  ): Promise<TxHashResponse> {
    return this.http.post<TxHashResponse>(
      `/api/v1/contracts/${enc(contractId)}/milestones/${enc(milestoneId)}/actions/reject`,
      params,
      opts,
    );
  }

  // -------------------------------------------------------------------------
  // Disputes
  // -------------------------------------------------------------------------

  /** Open a dispute on a contract. */
  async openDispute(
    contractId: string,
    params: ContractDisputeParams,
    opts?: RequestOptions,
  ): Promise<TxHashResponse> {
    return this.http.post<TxHashResponse>(
      `/api/v1/contracts/${enc(contractId)}/actions/dispute`,
      params,
      opts,
    );
  }

  /** Resolve a contract dispute. */
  async resolveDispute(
    contractId: string,
    params: ContractDisputeResolveParams,
    opts?: RequestOptions,
  ): Promise<TxHashResponse> {
    return this.http.post<TxHashResponse>(
      `/api/v1/contracts/${enc(contractId)}/actions/resolve`,
      params,
      opts,
    );
  }

  // -------------------------------------------------------------------------
  // Settlement
  // -------------------------------------------------------------------------

  /** Execute settlement (final payout). */
  async settlement(
    contractId: string,
    params: ContractSettlementParams,
    opts?: RequestOptions,
  ): Promise<TxHashResponse> {
    return this.http.post<TxHashResponse>(
      `/api/v1/contracts/${enc(contractId)}/actions/terminate`,
      params,
      opts,
    );
  }
}

function enc(s: string): string {
  return encodeURIComponent(s);
}
